File: app/src/main/java/network/minter/bipwallet/advanced/models/AddressListBalancesTotal.java
/*
 * Copyright (C) by MinterTeam. 2020
 * @link <a href="https://github.com/MinterTeam">Org Github</a>
 * @link <a href="https://github.com/edwardstock">Maintainer Github</a>
 *
 * The MIT License
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package network.minter.bipwallet.advanced.models;

import com.annimon.stream.Optional;
import com.annimon.stream.Stream;

import org.parceler.Parcel;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

import javax.annotation.Nonnull;

import network.minter.core.crypto.MinterAddress;
import network.minter.explorer.models.AddressBalance;

@Parcel
public class AddressListBalancesTotal {
    public List<AddressBalanceTotal> balances = new ArrayList<>();

    public AddressListBalancesTotal(List<MinterAddress> addresses) {
        for (MinterAddress address : addresses) {
            balances.add(new AddressBalanceTotal(address));
        }
    }

    public AddressListBalancesTotal() {
    }

    public int size() {
        return balances.size();
    }

    public AddressBalance get(int i) {
        return balances.get(i);
    }

    public Optional<AddressBalanceTotal> find(MinterAddress address) {
        return Stream.of(balances)
                .filter(item -> item.address != null)
                .filter(item -> item.address.equals(address))
                .findFirst();
    }

    public boolean isEmpty() {
        return size() == 0;
    }

    @Nonnull
    public AddressBalanceTotal getBalance(MinterAddress address) {
        for (AddressBalanceTotal balance : balances) {
            if (balance.address.equals(address)) {
                return balance;
            }
        }

        AddressBalanceTotal balance = new AddressBalanceTotal();
        balance.address = address;
        balance.fillDefaultsOnEmpty();
        return balance;
    }

    @Nonnull
    public BigDecimal getCoinBalance(MinterAddress address, String coin) {
        return getBalance(address).getCoin(coin).getAmount();
    }

}

File: app/src/main/java/network/minter/bipwallet/internal/storage/models/AddressListBalancesTotal.kt
/*
 * Copyright (C) by MinterTeam. 2022
 * @link <a href="https://github.com/MinterTeam">Org Github</a>
 * @link <a href="https://github.com/edwardstock">Maintainer Github</a>
 *
 * The MIT License
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package network.minter.bipwallet.internal.storage.models

import network.minter.bipwallet.internal.helpers.data.CollectionsHelper
import network.minter.core.crypto.MinterAddress
import network.minter.explorer.models.AddressBalance
import network.minter.explorer.models.CoinItemBase
import java.math.BigDecimal
import java.util.*

class AddressListBalancesTotal(
        var balances: MutableList<AddressBalanceTotal> = ArrayList(),
        var latestBlockTime: Date? = null
) {

    constructor(addresses: List<MinterAddress>) : this() {
        for (address in addresses) {
            balances.add(AddressBalanceTotal(address))
        }
    }

    fun size(): Int {
        return balances.size
    }

    operator fun get(i: Int): AddressBalance {
        return balances[i]
    }

    fun find(address: MinterAddress): AddressBalanceTotal? {
        return balances.firstOrNull { it.address == address }
    }

    fun remove(address: MinterAddress) {
        if (find(address) == null) {
            return
        }

        balances = CollectionsHelper.removeCopy(balances) {
            it.address == address
        }.toMutableList()
    }

    val isEmpty: Boolean
        get() = size() == 0

    /**
     * Find balance by address. Never returns null. If address not found, function fill it with zero values
     */
    fun getBalance(address: MinterAddress): AddressBalanceTotal {
        for (balance in balances) {
            if (balance.address == address) {
                return balance
            }
        }
        val balance = AddressBalanceTotal()
        balance.address = address
        balance.fillDefaultsOnEmpty()
        return balance
    }

    fun getCoinBalance(address: MinterAddress, coin: String): BigDecimal {
        return getBalance(address).getCoin(coin)?.amount ?: BigDecimal.ZERO
    }

    fun getCoinBalance(address: MinterAddress, coin: CoinItemBase): BigDecimal {
        return getBalance(address).getCoin(coin.id).amount ?: BigDecimal.ZERO
    }
}